pallets/utxo/src/lib.rs

```diff
@@ -753,21 +807,90 @@ pub mod pallet {
             ));
         }
 
+        let sig = crypto::sr25519_sign(
+            SR25519,
+            &sp_core::sr25519::Public::from_h256(public),
+            &tx.encode(),
+        )
+        .unwrap();
+        for i in 0..tx.inputs.len() {
+            tx.inputs[i].witness = sig.0.to_vec();
+        }
+        // Success
+        spend::<T>(caller, &tx)?;
+
         // Save in Store
         <TokenList<T>>::mutate(|x| {
-            if x.iter().find(|&x| x.id == token_id).is_none() {
+            if x.iter().find(|&x| x.id() == instance.id()).is_none() {
                 x.push(instance.clone())
             } else {
                 panic!("the token has already existed with the same id")
             }
         });
+        Ok(*instance.id())
+    }
+
+    fn mint<T: Config>(
+        caller: &T::AccountId,
+        creator_pubkey: sp_core::sr25519::Public,
+        data_url: String,
+        data_hash: [u8; 32],
+    ) -> Result<TokenId, DispatchErrorWithPostInfo<PostDispatchInfo>> {
+        let (fee, inputs_hashes) = pick_utxo::<T>(caller, Mlt(100).to_munit());
+        ensure!(fee >= Mlt(100).to_munit(), Error::<T>::Unapproved);
+        ensure!(data_url.len() <= 50, Error::<T>::Unapproved);
+
+        let instance = TokenInstance::new_nft(
+            BlakeTwo256::hash_of(&data_hash),
+            data_hash,
+            data_url.clone(),
+            creator_pubkey,
+        );
+
+        let inputs_for_fee = inputs_hashes
+            .iter()
+            .filter_map(|x| <UtxoStore<T>>::get(&x))
+            .map(|output| TransactionInput::new_empty(BlakeTwo256::hash_of(&(&output, 0 as u64))))
+            .collect();
+
+        ensure!(
+            !Nft::<T>::contains_key(instance.id()),
+            Error::<T>::NftCollectionExists
+        );
+
+        let mut tx = Transaction {
+            inputs: inputs_for_fee,
+            outputs: crate::vec![
+                // Output a new tokens
+                TransactionOutput::new_nft(
+                    *instance.id(),
+                    data_hash,
+                    data_url,
+                    H256::from(creator_pubkey)
+                ),
+            ],
+        };
 
+        let sig = crypto::sr25519_sign(SR25519, &creator_pubkey, &tx.encode()).unwrap();
```

Comment:
Does this work if you mint using Polkadot UI? It didn't work for me with `send_to_address()` but now I'm no longer sure should I remove that logic given that NFTs also depend on it. I was anyway going to discuss this with @muursh anyway but any thoughts regarding this?

Reply:
This will work if the node has access to the sk, as in the sk is stored using substrate's shitty km. Well it should work if that's the case. Does it fail if you try and sign something using the Alice's data which should be in the node's km? If it does do you have anything you can throw at me error wise?

Reply:
Another `unwrap()`... please stop using it... this is a disaster waiting to happen.

Reply:
IIRC it panicked or something similar at least when executed using a pallet call, I guess because it didn't have an access to the sk. I wonder why it works here but not with a pallet call...

Reply:
It may or may not work... my thesis is that unless the developer can **prove** that it'll always succeed, we should never see `unwrap()` in the code.

Reply:
@muursh @altonen Unfortunately, it took time to double-check, and I can't answer yet.

Reply:
So the default should be the tx being signed in the wallet, the average person won't be running a node with their sk stored in it. Signing using a locally stored sk is arguably an edge case. If the user tries it without a sk in km then that seems like a them problem to me

Reply:
@muursh  so the idea is that wallet only calls the `spend` pallet function which a signed transaction and possibly some other function to fetch the UTXOs of a given public key so that it can construct a TX? And those are the only two functions that wallet uses to interact with the UTXO system?